cmd/artifactPrepareVersion.go

```diff
@@ -0,0 +1,351 @@
+	if err != nil {
+		return errors.Wrap(err, "failed to initialize worktree")
+	}
+
+	return nil
+}
+
+func pushChanges(config *artifactPrepareVersionOptions, newVersion string, repository gitRepository, worktree gitWorktree, t time.Time) (string, error) {
+
+	var commitID string
+
+	commit, err := addAndCommit(worktree, newVersion, t)
+	if err != nil {
+		return commit.String(), err
+	}
+
+	commitID = commit.String()
+
+	tag := fmt.Sprintf("%v%v", config.TagPrefix, newVersion)
+	_, err = repository.CreateTag(tag, commit, nil)
+	if err != nil {
+		return commitID, err
+	}
+
+	ref := gitConfig.RefSpec(fmt.Sprintf("refs/tags/%v:refs/tags/%v", tag, tag))
+
+	pushOptions := git.PushOptions{
+		RefSpecs: []gitConfig.RefSpec{gitConfig.RefSpec(ref)},
+	}
+
+	currentRemoteOrigin, err := repository.Remote("origin")
+	if err != nil {
+		return commitID, errors.Wrap(err, "failed to retrieve current remote origin")
+	}
+	var updatedRemoteOrigin *git.Remote
+
+	urls := originUrls(repository)
+	if len(urls) == 0 {
+		return commitID, fmt.Errorf("no remote url maintained")
+	}
+	if strings.HasPrefix(urls[0], "http") {
+		if len(config.Username) == 0 || len(config.Password) == 0 {
+			// handling compatibility: try to use ssh in case no credentials are available
+			log.Entry().Info("git username/password missing - switching to ssh")
+
+			remoteURL := convertHTTPToSSHURL(urls[0])
+
+			// update remote origin url to point to ssh url instead of http(s) url
+			err = repository.DeleteRemote("origin")
+			if err != nil {
+				return commitID, errors.Wrap(err, "failed to update remote origin - remove")
+			}
+			updatedRemoteOrigin, err = repository.CreateRemote(&gitConfig.RemoteConfig{Name: "origin", URLs: []string{remoteURL}})
+			if err != nil {
+				return commitID, errors.Wrap(err, "failed to update remote origin - create")
+			}
+
+			pushOptions.Auth, err = sshAgentAuth("git")
+			if err != nil {
+				return commitID, errors.Wrap(err, "failed to retrieve ssh authentication")
+			}
+			log.Entry().Infof("using remote '%v'", remoteURL)
+		} else {
+			pushOptions.Auth = &http.BasicAuth{Username: config.Username, Password: config.Password}
+		}
+	} else {
+		pushOptions.Auth, err = sshAgentAuth("git")
+		if err != nil {
+			return commitID, errors.Wrap(err, "failed to retrieve ssh authentication")
+		}
+	}
+
+	err = repository.Push(&pushOptions)
+	if err != nil {
+		return commitID, err
+	}
+
+	if updatedRemoteOrigin != currentRemoteOrigin {
+		err = repository.DeleteRemote("origin")
+		if err != nil {
+			return commitID, errors.Wrap(err, "failed to restore remote origin - remove")
+		}
+		_, err := repository.CreateRemote(currentRemoteOrigin.Config())
+		if err != nil {
+			return commitID, errors.Wrap(err, "failed to restore remote origin - create")
+		}
+	}
+
+	return commitID, nil
+}
+
+func addAndCommit(worktree gitWorktree, newVersion string, t time.Time) (plumbing.Hash, error) {
+	_, err := worktree.Add(".")
```

Comment:
this should use the `update` flag like here https://github.com/SAP/jenkins-library/blob/master/vars/artifactSetVersion.groovy#L194 to avoid adding files that should not be commited

Reply:
The command here does not have a flag `update`. Let's see if there is a need for further additions further down the road.

Reply:
We need this to avoid committing unneeded files in the sdk pipeline, that is why I added it in the first place.

Reply:
Also it seems like gitignore is ignored, so using this locally in a checkout with target directories etc results in a large commit with all binary files. this is not the expected behaviour.